### PR TITLE
feat(core/date-/datetime-input): Add new components ix-date-input and ix-datetime-input

### PR DESCRIPTION
**Pull request by lzeiml**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Unit tests (`yarn test`) were run locally and passed
- [x] Visual Regression Tests (`yarn visual-regression`) were run locally and passed
- [x] Linting (`npm lint`) was run locally and passed

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bug fix
- [x] Feature
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the new behavior?

- New ix-date-input component that uses the ix-date-picker-rework
- New ix-datetime-input component that uses the ix-datetime-picker-rework

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

Added component tests and snapshot tests for both components

### CODE REVIEW

File: packages/core/src/components/date-input/date-input.tsx [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+          </span>
+        </div>
+        <ix-dropdown
+          trigger={this.dateInputDiv}
+          closeBehavior="outside"
+          onClick={(event) => event.stopPropagation()}
+          class="dropdown"
+        >
+          <ix-date-picker-rework
+            tabIndex={0}
```
Comment (nuke-ellington) [inline packages/core/src/components/date-input/date-input.tsx:438]: The picker shows the default white border on focus. Ideally the focus would jump to an interactive element directly.

**Review by nuke-ellington — CHANGES_REQUESTED** [review]
Good stuff. Here are some minor change requests.
Additionally these points should be adressed:

disabled state/property for new components
a11y support as specified [here](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/datepicker-dialog/)

**Comment by nuke-ellington** [comment]
Adresses second bullet point from #384

File: packages/angular-test-app/src/app/app.component.ts
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/angular-test-app/src/app/app.component.ts:1]: https://github.com/siemens/ix/pull/861/commits/5a56d5b52bb924eaeabaed137dfd6fed5448325a

File: packages/core/src/components/date-input/date-input.scss [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+  span {
+    &.vertical-align {
+      display: flex;
+      flex-direction: row;
+      align-items: center;
+    }
+
+    &.left-position {
+      margin-right: 10px;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.scss]: https://github.com/siemens/ix/pull/861/commits/7eeb8efa55800ce6fa5ddf86832cc9bbee35f7c7

File: packages/core/src/components/date-input/date-input.tsx [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+  ValidatorName,
+} from '../utils/validators/validator.factory';
+
+export interface DateInputEvent {
+  InputName: string;
+  Event: Event;
+}
+
+/**
+ * @since 3.0.0
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.tsx]: https://github.com/siemens/ix/pull/861/commits/8eeb26d6ed7c1a34105492ce3e2712170fb5211a

File: packages/core/src/components/date-input/date-input.tsx [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+  @Prop() labelPosition: 'top' | 'left' | 'inside' = 'top';
+
+  /**
+   * Date format string.
+   * See {@link "https://day.js.org/docs/en/display/format"} for all available tokens.
+   */
+  @Prop() format: string = 'YYYY/MM/DD';
+
+  /**
+   * If true a date-range can be selected (from/to).
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.tsx]: https://github.com/siemens/ix/pull/861/commits/5d0d6123441665ee65ddd8fc48a3e1db687638de

File: packages/core/src/components/date-input/date-input.tsx [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+  /**
+   * The latest date that can be selected by the date picker.
+   * If not set there will be no restriction.
+   */
+  @Prop() maxDate: string;
+
+  /**
+   * Text of the button that confirms date selection.
+   */
+  @Prop() textSelectDate = 'Done';
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.tsx]: https://github.com/siemens/ix/pull/861/commits/a140f27cc1865ac37d810f00fdc1b918c3308dca

File: packages/core/src/components/date-input/date-input.tsx [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+  /**
+   * The index of which day to start the week on, based on the Locale#weekdays array.
+   * E.g. if the locale is en-us, weekStartIndex = 1 results in starting the week on monday.
+   */
+  @Prop() weekStartIndex = 0;
+
+  /**
+   * Suppresses the error message that displays below the input if validation failed.
+   */
+  @Prop() suppressErrorHandlingMessage: boolean = false;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.tsx]: https://github.com/siemens/ix/pull/861/commits/a140f27cc1865ac37d810f00fdc1b918c3308dca

File: packages/core/src/components/date-input/date-input.tsx [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+   */
+  @Prop() validators: InputValidator[] | string[] = [
+    'validDate',
+    'toAfterFrom',
+    'withinMinMax',
+  ];
+
+  /**
+   * Triggers if the date selection changes.
+
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.tsx]: https://github.com/siemens/ix/pull/861/commits/79ed2d955c3c7b246097a33d3aa5b3523f40493b

File: packages/core/src/components/date-input/date-input.tsx [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+   * Triggers if one of the inputs changes
+   * @emits DateInputEvent
+   */
+  @Event() ixOnChange: EventEmitter<DateInputEvent>;
+
+  /**
+   * Triggers if one of the inputs gets focus
+   * @emits DateInputEvent
+   */
+  @Event() ixOnFocus: EventEmitter<DateInputEvent>;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.tsx]: https://github.com/siemens/ix/pull/861/commits/79ed2d955c3c7b246097a33d3aa5b3523f40493b

File: packages/core/src/components/date-input/date-input.tsx [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+
+  /**
+   * Triggers if one of the inputs loses focus
+   * @emits DateInputEvent
+   */
+  @Event() ixOnBlur: EventEmitter<DateInputEvent>;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.tsx]: https://github.com/siemens/ix/pull/861/commits/79ed2d955c3c7b246097a33d3aa5b3523f40493b

File: packages/core/src/components/date-input/date-input.tsx [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+
+  /**
+   * Triggers if the inputs get cleared by pressing the clear button
+   * @emits void
+   */
+  @Event() ixOnClear: EventEmitter<void>;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.tsx]: https://github.com/siemens/ix/pull/861/commits/79ed2d955c3c7b246097a33d3aa5b3523f40493b

File: packages/core/src/components/date-input/date-input.tsx [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+  /**
+   * Date selection confirmed via button action
+   */
+  @Event() dateSelect: EventEmitter<DateChangeEventRework>;
+
+  /**
+   * Triggers if one of the inputs changes
+   * @emits DateInputEvent
+   */
+  @Event() ixOnChange: EventEmitter<DateInputEvent>;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.tsx]: https://github.com/siemens/ix/pull/861/commits/79ed2d955c3c7b246097a33d3aa5b3523f40493b

File: packages/core/src/components/datetime-input/datetime-input.scss [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    flex-direction: row;
+    width: 100%;
+
+    .range-input {
+      flex: 0 0 50%;
+    }
+  }
+
+  .second-input {
+    margin-left: 10px;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/datetime-input/datetime-input.scss]: https://github.com/siemens/ix/pull/861/commits/f44e26cce3818678daabafa91777d135ee6217e9

File: packages/core/src/components/datetime-input/datetime-input.tsx [hunk 13]
```diff
@@ -0,0 +1,682 @@
...
+type DateChangeCallback = (
+  event: IxDatetimePickerReworkCustomEvent<DateTimeDateChangeEvent>
+) => void;
+
+type TimeChangeCallback = (
+  event: IxDatetimePickerReworkCustomEvent<string>
+) => void;
+
+/**
+ * @since 3.0.0
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/datetime-input/datetime-input.tsx]: https://github.com/siemens/ix/pull/861/commits/b72916289ed08bbb65947e3131e3970cad092314

File: packages/core/src/components/datetime-input/datetime-input.tsx [hunk 14]
```diff
@@ -0,0 +1,682 @@
...
+   */
+  @Prop() label: string;
+
+  /**
+   * Position of the label
+   */
+  @Prop() labelPosition: 'above' | 'inline' = 'above';
+
+  /**
+   * Set range size
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/datetime-input/datetime-input.tsx:83]: https://github.com/siemens/ix/pull/861/commits/b72916289ed08bbb65947e3131e3970cad092314

File: packages/core/src/components/datetime-input/datetime-input.tsx [hunk 15]
```diff
@@ -0,0 +1,682 @@
...
+
+  /**
+   * Set time reference
+   */
+  @Prop() timeReference: 'AM' | 'PM';
+
+  /**
+   * Text of date select button
+   */
+  @Prop() textSelectDate = 'Done';
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/datetime-input/datetime-input.tsx]: https://github.com/siemens/ix/pull/861/commits/a140f27cc1865ac37d810f00fdc1b918c3308dca

File: packages/core/src/components/datetime-input/datetime-input.tsx [hunk 16]
```diff
@@ -0,0 +1,682 @@
...
+   * @emits DateTimeSelectEvent
+   */
+  @Event() dateSelect: EventEmitter<DateTimeSelectEvent>;
+
+  /**
+   * Triggers every time one of the inputs changes
+   *
+   * @emits DatetimeInputChangeEvent
+   */
+  @Event() ixOnChange: EventEmitter<DatetimeInputEvent>;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/datetime-input/datetime-input.tsx]: https://github.com/siemens/ix/pull/861/commits/79ed2d955c3c7b246097a33d3aa5b3523f40493b

File: packages/core/src/components/datetime-input/datetime-input.tsx [hunk 17]
```diff
@@ -0,0 +1,682 @@
...
+
+  /**
+   * Triggers if one of the inputs gets focus
+   * @emits DatetimeInputEvent
+   */
+  @Event() ixOnFocus: EventEmitter<DatetimeInputEvent>;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/datetime-input/datetime-input.tsx]: https://github.com/siemens/ix/pull/861/commits/79ed2d955c3c7b246097a33d3aa5b3523f40493b

File: packages/core/src/components/datetime-input/datetime-input.tsx [hunk 18]
```diff
@@ -0,0 +1,682 @@
...
+
+  /**
+   * Triggers if one of the inputs loses focus
+   * @emits DatetimeInputEvent
+   */
+  @Event() ixOnBlur: EventEmitter<DatetimeInputEvent>;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/datetime-input/datetime-input.tsx]: https://github.com/siemens/ix/pull/861/commits/79ed2d955c3c7b246097a33d3aa5b3523f40493b

File: packages/core/src/components/datetime-input/datetime-input.tsx [hunk 19]
```diff
@@ -0,0 +1,682 @@
...
+
+  /**
+   * Triggers if the inputs get cleared by pressing the clear button
+   * @emits string with the name of the input that was cleared
+   */
+  @Event() ixOnClear: EventEmitter<string>;
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/datetime-input/datetime-input.tsx]: https://github.com/siemens/ix/pull/861/commits/79ed2d955c3c7b246097a33d3aa5b3523f40493b

File: packages/core/src/components/datetime-input/datetime-input.tsx [hunk 20]
```diff
@@ -0,0 +1,682 @@
...
+            </span>
+            <span class="icon-button">
+              <ix-icon-button ghost icon="chevron-down-small"></ix-icon-button>
+            </span>
+          </div>
+        </div>
+        <ix-dropdown
+          trigger={isSecondInput ? this.toInputDiv : this.fromInputDiv}
+          closeBehavior="outside"
+          // onClick={(event) => event.stopPropagation()}
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/core/src/components/datetime-input/datetime-input.tsx]: https://github.com/siemens/ix/pull/861/commits/e4208c2b72b25c5245c57efdbd6db48a0633bd48

File: packages/documentation/docs/controls/date-time-picker.md [hunk 21]
```diff
@@ -7,7 +7,7 @@ import SourceReactDatetimepicker from './../auto-generated/previews/react/dateti
 import SourceAngularDatetimepicker from './../auto-generated/previews/angular/datetimepicker.ts.md'
 import SourceVueDatetimepicker from './../auto-generated/previews/vue/datetimepicker.md'
 
-# Date time picker
+# Datetime picker
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/documentation/docs/controls/date-time-picker.md]: https://github.com/siemens/ix/pull/861/commits/043724f2dc881bf0d398a81e21468dfb8ae574ee

File: packages/documentation/docs/controls/datetime-input.md [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+import SourceDatetimeInputRange from './../auto-generated/previews/web-component/datetime-input-range.md'
+import SourceReactDatetimeInputRange from './../auto-generated/previews/react/datetime-input-range.md'
+import SourceAngularDatetimeInputRange from './../auto-generated/previews/angular/datetime-input-range.ts.md'
+import SourceVueDatetimeInputRange from './../auto-generated/previews/vue/datetime-input-range.md'
+
+import SourceDatetimeValidation from './../auto-generated/previews/web-component/datetime-input-validation.md'
+import SourceReactDatetimeValidation from './../auto-generated/previews/react/datetime-input-validation.md'
+import SourceAngularDatetimeValidation from './../auto-generated/previews/angular/datetime-input-validation.ts.md'
+
+# Datetime input
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/documentation/docs/controls/datetime-input.md]: https://github.com/siemens/ix/pull/861/commits/043724f2dc881bf0d398a81e21468dfb8ae574ee

File: packages/documentation/docs/controls/datetime-input.md [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+## Range Selection
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/documentation/docs/controls/datetime-input.md]: https://github.com/siemens/ix/pull/861/commits/043724f2dc881bf0d398a81e21468dfb8ae574ee

File: packages/documentation/docs/controls/datetime-input.md [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+name="datetime-input-range" height="35rem"
+hideInitalCodePreview
+frameworks={{
+  react: SourceReactDatetimeInputRange,
+  angular: SourceAngularDatetimeInputRange,
+  javascript: SourceDatetimeInputRange,
+  vue: SourceVueDatetimeInputRange
+}}></Playground>
+
+## Single Selection
```
[the remark this answers is not in the document]
Comment (lzeiml) [reply packages/documentation/docs/controls/datetime-input.md]: https://github.com/siemens/ix/pull/861/commits/043724f2dc881bf0d398a81e21468dfb8ae574ee

File: packages/core/src/components/date-input/date-input.tsx (hunk 1 above)
Comment (lzeiml) [reply packages/core/src/components/date-input/date-input.tsx]: https://github.com/siemens/ix/pull/861/commits/e98290a4da66a99f43fa1d92ee9796abc6bc7f13

**Comment by lzeiml** [comment]
>a11y support

https://github.com/siemens/ix/pull/861/commits/6e53d8156018a691a86620967f81d349c71542c9

**Comment by lzeiml** [comment]
> disabled state/property for new components
https://github.com/siemens/ix/pull/861/commits/c2ba2a6dda1492fa19c66b6046c030aaa86ff1a4

File: packages/core/src/tests/date-input/date-input.e2e.ts-snapshots/date-input-range-1-chromium---theme-classic-dark-linux.png
[the remark this answers is not in the document]
Comment (nuke-ellington) [reply packages/core/src/tests/date-input/date-input.e2e.ts-snapshots/date-input-range-1-chromium---theme-classic-dark-linux.png:1]: Looking at Figma there seems to be too much margin around the arrow still

File: packages/core/src/tests/datetime-input/datetime-input.e2e.ts-snapshots/datetime-input-range-1-chromium---theme-classic-dark-linux.png
[the remark this answers is not in the document]
Comment (nuke-ellington) [reply packages/core/src/tests/datetime-input/datetime-input.e2e.ts-snapshots/datetime-input-range-1-chromium---theme-classic-dark-linux.png:1]: Also still too much margin between the input fields

File: packages/core/src/components/datetime-input/datetime-input.tsx [hunk 25]
```diff
@@ -0,0 +1,749 @@
...
+
+  /**
+   * Position of the label
+   */
+  @Prop() labelPosition: 'top' | 'left' | 'inside' = 'top';
+
+  /**
+   * Set range size
+   */
+  @Prop() range = true;
```
Comment (nuke-ellington) [inline packages/core/src/components/datetime-input/datetime-input.tsx:85]: Maybe default should be `false`? TBD with UX team

**Review by nuke-ellington — CHANGES_REQUESTED** [review]
Some minor changes plus:

- VRT for disabled state is missing
- VRT for label positions are missing

**Comment by nuke-ellington** [comment]
[IX-641] has some UX TBDs that need to be clarified before this can be finalized

**Comment by lzeiml** [comment]
>This fix does not adress this JSDoc at all. Am I missing something?

https://github.com/siemens/ix/pull/861/commits/9cc640e74b53051ef73cbd6aec9120dff3ad28d5

**Comment by lzeiml** [comment]
> Looking at Figma there seems to be too much margin around the arrow still

I don't know how to implement 1:1 representation of the figma spec. I am setting the width of the inputs based on the placeholder attribute, which is usually wider than an entered value, which makes it appear to have more margin.

https://github.com/siemens/ix/pull/861/commits/5af1fea38ed16dd47848a314656839e656fcbbac

**Comment by lzeiml** [comment]
> VRT for disabled state is missing
> VRT for label positions are missing

https://github.com/siemens/ix/pull/861/commits/e3c4d4f16c721a270eceb9fe427ca58651321bb3

File: packages/core/src/components/datetime-input/datetime-input.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (nuke-ellington) [reply packages/core/src/components/datetime-input/datetime-input.tsx:83]: I think this was a missunderstanding. The JSDocs for range are not very selfexplanatory

File: packages/core/src/components/date-input/date-input.tsx [hunk 26]
```diff
@@ -0,0 +1,491 @@
...
+
+    hiddenInput.form.addEventListener('submit', this.setInputValidity);
+  }
+
+  componentDidRender() {
+    const inputs = this.hostElement.shadowRoot.querySelectorAll('input');
+    inputs.forEach((input) => {
+      input.setAttribute(
+        'size',
+        input.getAttribute('placeholder').length as unknown as string
```
Comment (nuke-ellington) [inline packages/core/src/components/date-input/date-input.tsx:341]: I think this approach is fine but we should factor in the input padding so nothing gets cut off.

**Comment by Aiderlei** [comment]
When is this going to be released?

**Comment by danielleroux** [comment]
Will be implemented with #1318

**Pull request closed without merging**